day3/m_partitioner.F90

```diff
@@ -98,17 +107,50 @@ type(Partitioner) function constructor(comm, domain, num_halo, periodic) result(
     this%global_shape_(2) = domain(2) + 2 * num_halo
     this%global_shape_(3) = domain(3)
 
-    size = this%setup_grid()
+    call this%setup_grid()
     call this%setup_domain(domain, num_halo)
-  end function
+
+    if (debug) then
+      call MPI_Barrier(this%comm_, ierror)
+
+      if (this%rank_ == 0) then
+        write(error_unit, '(a33, /, a30, 3(i6), /, a30, 1(i6), /, a30, 2(l6))') &
+          '====== Global information =======', &
+          'Domain size (nx,ny,nz): ', this%global_shape_ - [2 * num_halo, 2 * num_halo, 0], &
+          'Halo width: ', this%num_halo_, &
+          'Periodicity (x,y): ', this%periodic_
+      end if
+
+      do rank = 0, this%num_ranks_ - 1
+        call MPI_Barrier(this%comm_, ierror)
+        if (this%rank_ == rank) then
+          call error(ierror /= MPI_SUCCESS, 'Problem with MPI_Barrier', code = ierror)
+          write(error_unit, '(a12, i13, a8, /, a30, 2(i6), /, a30, 3(i6), /, a30, 4(i6), / a30, 4(i6))') &
+          '====== Rank ', rank, ' =======', &
+          'Position (x,y): ', this%position(), &
+          'Subdomain size (nx,ny,nz): ', this%shape() - [2 * num_halo, 2 * num_halo, 0], &
+          'Position on global grid: ', this%compute_domain(), &
+          'Neighbors (trbl): ', this%top(), this%right(), this%bottom(), this%left()
+        end if
+        call MPI_Barrier(this%comm_, ierror)
+      end do
+
+      if (this%rank_ == 0) then
+        write(error_unit, '(a33)') '================================='
+      end if
+
+      call MPI_Barrier(this%comm_, ierror)
```

Comment:
`stderr`, ie. `error_unit` is not buffered, therefore `flush` is not needed here. OTOH multiple `write`s can be interspersed between ranks, but single `write`s are atomic, so I folded them

Reply:
Unfortunately, same issue as above with `include mpi`... reality shows that several MPI implementations don't sync stdout/stderr between ranks unless flush is inserted. See for example https://www-uxsup.csx.cam.ac.uk/courses/moved.MPI/paper_08.pdf or https://www.archer.ac.uk/training/course-material/2017/09/mpi-york/slides/L03-archer-cirrus-mpi.pdf or https://software.intel.com/en-us/forums/intel-clusters-and-hpc-technology/topic/332514